### PR TITLE
Performance - 8x Faster TryConvertTo

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
I was running some benchmarking in my own mapping library and I realised,after looking at the Umbraco source, I could dramatically improve the performance of the `TryConvertTo` extension method variants which, since, they are so commonly used, should give a nice overall performance boost to Umbraco.

The PR contains improvements to those core methods gained by various optimization techniques.
I also updated the `Umbraco.Tests.Benchmarks` library to NET 4.6 so that we could update to the latest BenchmarkDotNet version (0.10.12 at this time), cleaned up the test configuration, and added benchmarks for the improved features.

As you can see, in the following environment most of the scenarios are **~8x faster** consuming less memory with a slight improvement in the already well optimized `DateTime` conversions also.

All underlying logic follows the same principle as before (same TODO:s etc), only optimizations were applied.

#### Benchmarks

```
BenchmarkDotNet=v0.10.12, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.192)
Intel Core i7-6600U CPU 2.60GHz (Skylake), 1 CPU, 4 logical cores and 2 physical cores
Frequency=2742184 Hz, Resolution=364.6728 ns, Timer=TSC
  [Host]     : .NET Framework 4.7 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.2600.0
  DefaultJob : .NET Framework 4.7 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.2600.0
```

#### Before:

Method |     Mean |     Error |    StdDev |  Gen 0 | Allocated |
-------------------------------------- |---------:|----------:|----------:|-------:|----------:|
'List<string> to IEnumerable<string>' | 2.545 us | 0.0496 us | 0.0712 us | 0.0420 |      96 B |
'Int to Double' | 2.657 us | 0.0461 us | 0.0431 us | 0.0458 |     104 B |
'Float to Decimal' | 3.259 us | 0.0794 us | 0.1005 us | 0.0610 |     132 B |
'String to Boolean' | 2.238 us | 0.0457 us | 0.0507 us | 0.0458 |     100 B |
'String to DateTime' | 1.936 us | 0.0378 us | 0.0435 us | 0.0038 |      16 B |

#### After:

Method |       Mean |     Error |    StdDev |  Gen 0 | Allocated |
-------------------------------------- |-----------:|----------:|----------:|-------:|----------:|
'List<string> to IEnumerable<string>' |   361.3 ns |  7.077 ns |  7.268 ns |      - |       0 B |
'Int to Double' |   251.6 ns |  5.235 ns |  7.165 ns | 0.0186 |      40 B |
'Float to Decimal' |   382.7 ns |  7.552 ns | 11.979 ns | 0.0172 |      36 B |
'String to Boolean' |   288.5 ns |  5.586 ns |  7.065 ns | 0.0052 |      12 B |
'String to DateTime' | 1,819.3 ns | 27.798 ns | 26.002 ns | 0.0057 |      16 B |

I don't know what the process is for submitting to V8 (do improvements her get merged across?) but can submit a similar PR there if required.

### CODE REVIEW

**Comment by madsoulswe** [comment]
8x...
giphy 1

**Comment by zpqrtbnk** [comment]
The string helper (replace many) thing is weird, I clearly remember running benchmarks but can not find them anymore, and obviously now your way is fastest - approved!

Conversion review in progress...

**Comment by JimBobSquarePants** [comment]
I've been tripped over before running benchmarks with and without my laptop plugged in. The different power modes dramatically affect results.

**Comment by zpqrtbnk** [comment]
Impressed ;-)

Noting that none of the cache dictionaries are concurrent - pro/cons?

**Comment by antonfirsov** [comment]
I have no knowledge of Umbraco codebase, but if it's a concurrent environment, these methods should be probably thread safe.
Could be done with `ConcurrentDictionary`, but it has some overheads.

**Comment by JimBobSquarePants** [comment]
There's no risk here that requires the overhead of a concurrent lookup. If we miss the cache we simply reassign the same viable again via the indexer. That's why I chose to add the items that way.

**Comment by antonfirsov** [comment]
The problem is that the indexer is not atomic. If two threads happen [to write to the cache](https://github.com/JimBobSquarePants/Umbraco-CMS/blob/24071eb3b1294461fde8c0c9d1502135d9efbfff/src/Umbraco.Core/ObjectExtensions.cs#L662) simultaneously it can corrupt the internal state of `Dictionary<K,V>`.

I haven't done deep analysis on the code however, so maybe I'm just missing something?

**Comment by zpqrtbnk** [comment]
I probably need to read the doc again, but my understanding was that concurrent accesses to a `Dictionary` were not safe: reading right when another thread writes, or concurrent writes... can cause weird things to happen. Would tend to prefer to use concurrent dictionaries. Thoughts?

**Comment by zpqrtbnk** [comment]
Have merged the PR into branch https://github.com/umbraco/Umbraco-CMS/tree/temp-james-perfs and pushed an additional cleanup commit on top of it, mostly rearranging stuff.

Would love to get the dictionary thing sorted w/James then could be merged.

As for v8, yes all changes to v7 are currently incrementally merged into v8, so that would not be lost.

**Comment by JimBobSquarePants** [comment]
We should switch them out for ConcurrentDictionary. I forgot about when the dictionary resizes. That could potentially cause a locked thread. Should still be an order of magnitude faster.

**Comment by JimBobSquarePants** [comment]
Confirmed locally. Combined with the better use of `TryGetValue` that you've introduced there's no perceivable difference in performance and you get the thread safety. I'll let you make the changes now as I have no idea what to push to. (Normally changes are made to incoming PR's, I checked the box to allow that).

**Comment by zpqrtbnk** [comment]
Oh I can push to your PR? Silly me did not even try.
Bad Stephan.

OK, done now, so we can resume work on this very PR.
Will keep the other branch on Core just long enough to read your comments.

Oh and Slack tells me you've written "*I really don't like this; you cannot possibly know the type at this point. I honestly find these kind of rules off-putting to contribute code to.*" as a comment at some point but I cannot find it anywhere in GitHub. Can you tell me more about it?

**Comment by zpqrtbnk** [comment]
And then, have pushed a commit moving from `Dictionary` to `ConcurrentDictionary` - but you'll find a comment noted `// FIXME` in there - not sure I understand `GetCachedAssignableConvertibleResult` - we are caching something that does not depend on the key?

**Comment by zpqrtbnk** [comment]
Finally, pushed a commit dealing with the `// FIXME` I mention above.

**Comment by JimBobSquarePants** [comment]
Thanks for the update. 

Re the comment. I have a bee in my bonnet regarding `var` when you can't see the type on the right hand side as I feel it reduces readability. I deleted the comment though as it's your game with your rules. 

Using the `bool` return type is a good update (my code was the leftover of previous logic) though `GetOrAdd` is slower than the `TryGetValue` logic you had previously due to the overhead of `Func.Invoke`.

**Comment by zpqrtbnk** [comment]
And I made sure to not use captured vars in `GetOrAdd` to avoid allocs... Bah... Now I'm wondering... reverting to plain dictionaries and wrapping all accesses in a `lock` statement will improve the situation in benchmarks, I guess, but benchmarks don't deal with multiple concurrent readers. We're probably as fast-and-safe as we can now, what do you think?

File: src/Umbraco.Core/ObjectExtensions.cs [hunk 1]
```diff
@@ -652,5 +644,69 @@ internal static Guid AsGuid(this object value)
...
-	}
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static string NormalizeNumberDecimalSeparator(string s)
+        {
+            var normalized = System.Threading.Thread.CurrentThread.CurrentCulture.NumberFormat.NumberDecimalSeparator[0];
+            return s.ReplaceMany(NumberDecimalSeparatorsToNormalize, normalized);
+        }
+
+        // gets a converter for source, that can convert to target, or null if none exists
```
Comment (JimBobSquarePants) [inline src/Umbraco.Core/ObjectExtensions.cs]: If adding comments why not proper intellisense ones so you can see them from the calling code?

File: src/Umbraco.Core/ObjectExtensions.cs [hunk 2]
```diff
@@ -652,5 +644,69 @@ internal static Guid AsGuid(this object value)
...
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static TypeConverter GetCachedSourceTypeConverter(Type source, Type target)
+        {
+            return InputTypeConverterCache.GetOrAdd(new CompositeTypeTypeKey(source, target), k =>
+            {
+                var ksource = k.Type1;
```
Comment (JimBobSquarePants) [inline src/Umbraco.Core/ObjectExtensions.cs]: Even if you ensure there's no extra allocation here there's an extra null check against the factory method for every call when using the `GetOrAdd` convenience method. Calling `TryGetValue` directly in the following manner is the more efficient code. 

``` c#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static TypeConverter GetCachedSourceTypeConverter(Type source, Type target)
{
    var key = new CompositeTypeTypeKey(source, target);

    if (InputTypeConverterCache.TryGetValue(key, out TypeConverter typeConverter))
    {
        return typeConverter;
    }

    TypeConverter converter = TypeDescriptor.GetConverter(source);
    if (converter.CanConvertTo(target))
    {
        return InputTypeConverterCache[key] = converter;
    }

    return InputTypeConverterCache[key] = null;
}
```

**Comment by JimBobSquarePants** [comment]
I wouldn't revert back to dictionaries now. We can gain a benefit from any future improvements to the Core Clr using their methods without risk of us making mistakes.

I'm gonna push a commit after work reverting `GetOrAdd` to `TryGetValue` in our caching methods since it's faster for all calls (see the code comment). That won't be for another 8-10 hours though.

**Comment by JimBobSquarePants** [comment]
Ok that's the fastest way to do the lookups in a threadsafe manner. This is still > 8x faster with all the trimmings :smile: 

shake-n-bake

File: src/Umbraco.Core/ObjectExtensions.cs (hunk 1 above)
Comment (zpqrtbnk) [reply src/Umbraco.Core/ObjectExtensions.cs]: Weird tradition of mine for entirely private methods. But I think you're right on that one.

**Comment by zpqrtbnk** [comment]
Looks quite good now. Still have one question regarding `GetCachedCanAssign` where I don't quite like that we pass `input` to the method and make a decision based on `input`, yet `input` is not part of the cache key. I *guess* it is, since `inputType` is the type of `input`, but I find it confusing. 

But maybe it is because `input is IConvertible` is faster than `typeof(IConvertible).IsAssignableFrom(ksource)`? If that's the reason then fine, but then I'll add some notes to explain, else someone else is going to clean it up later on.

**Comment by JimBobSquarePants** [comment]
>But maybe it is because input is IConvertible is faster than typeof(IConvertible).IsAssignableFrom(ksource)? If that's the reason then fine, but then I'll add some notes to explain, else someone else is going to clean it up later on.

Yeah, that's exactly why. We can store the input type in our key but that very fast `is` allows us to very quickly decide whether we should even try `IsAssignableFrom()`.  Maybe overkill, but I can't miss the opportunity to speed things up when I can.

I left a note in the benchmarks but yeah, probably wise to add one there also. I'll commit and do so.

I think I've done what I can with this now and will leave it up to you guys. Should be legit.

**Pull request merged**